File: scenarios/landing-page-business-suite/ui/src/shared/api/feedback.ts
import { createClient } from '@connectrpc/connect';
import { createScenarioConnectTransport } from '@vrooli/api-base';
import {
  FeedbackService,
  FeedbackStatus,
  FeedbackType,
  type FeedbackRecord as GeneratedFeedbackRecord,
} from '@vrooli/proto-types/landing-page-business-suite/feedback_pb';
import { CONNECT_API_BASE } from './common';

const feedbackClient = createClient(FeedbackService, createScenarioConnectTransport({ baseUrl: CONNECT_API_BASE }));

export interface FeedbackRequest {
  id: number;
  type: 'refund' | 'bug' | 'feature' | 'general';
  email: string;
  subject: string;
  message: string;
  order_id?: string | null;
  status: 'pending' | 'in_progress' | 'resolved' | 'rejected';
  created_at: string;
  updated_at: string;
}

export interface CreateFeedbackInput {
  type: string;
  email: string;
  subject: string;
  message: string;
  orderId?: string;
}

function typeFromProto(value: FeedbackType): FeedbackRequest['type'] {
  switch (value) {
    case FeedbackType.REFUND: return 'refund';
    case FeedbackType.BUG: return 'bug';
    case FeedbackType.FEATURE: return 'feature';
    default: return 'general';
  }
}

function statusFromProto(value: FeedbackStatus): FeedbackRequest['status'] {
  switch (value) {
    case FeedbackStatus.IN_PROGRESS: return 'in_progress';
    case FeedbackStatus.RESOLVED: return 'resolved';
    case FeedbackStatus.REJECTED: return 'rejected';
    default: return 'pending';
  }
}

function statusToProto(value?: string): FeedbackStatus | undefined {
  switch (value) {
    case undefined: return undefined;
    case 'pending': return FeedbackStatus.PENDING;
    case 'in_progress': return FeedbackStatus.IN_PROGRESS;
    case 'resolved': return FeedbackStatus.RESOLVED;
    case 'rejected': return FeedbackStatus.REJECTED;
    default: throw new Error(`Invalid feedback status: ${value}`);
  }
}

function timestampToISO(value: GeneratedFeedbackRecord['createdAt']): string {
  if (!value) return '';
  return new Date(Number(value.seconds) * 1000 + value.nanos / 1_000_000).toISOString().replace('.000Z', 'Z');
}

function feedbackFromProto(value: GeneratedFeedbackRecord): FeedbackRequest {
  return {
    id: Number(value.id), type: typeFromProto(value.type), email: value.email, subject: value.subject,
    message: value.message, ...(value.orderId !== undefined ? { order_id: value.orderId } : {}),
    status: statusFromProto(value.status), created_at: timestampToISO(value.createdAt), updated_at: timestampToISO(value.updatedAt),
  };
}

export async function createFeedback(input: CreateFeedbackInput, signal?: AbortSignal): Promise<{ success: boolean; id: number }> {
  const response = await feedbackClient.createFeedback(
    { type: input.type, email: input.email, subject: input.subject, message: input.message, ...(input.orderId !== undefined ? { orderId: input.orderId } : {}) },
    { signal },
  );
  return { success: response.success, id: Number(response.id) };
}

export async function fetchFeedbackList(status?: string): Promise<FeedbackRequest[]> {
  const response = await feedbackClient.listFeedback({ status: statusToProto(status) });
  return response.feedback.map(feedbackFromProto);
}

export async function fetchFeedbackById(id: number): Promise<FeedbackRequest> {
  const response = await feedbackClient.getFeedback({ id: BigInt(id) });
  if (!response.feedback) throw new Error('Feedback response did not include a record');
  return feedbackFromProto(response.feedback);
}

export async function updateFeedbackStatus(id: number, status: string): Promise<FeedbackRequest> {
  const protoStatus = statusToProto(status);
  if (protoStatus === undefined) throw new Error('Feedback status is required');
  const response = await feedbackClient.updateFeedbackStatus({ id: BigInt(id), status: protoStatus });
  if (!response.feedback) throw new Error('Feedback response did not include a record');
  return feedbackFromProto(response.feedback);
}

export async function deleteFeedback(id: number): Promise<{ success: boolean; id: number }> {
  const response = await feedbackClient.deleteFeedback({ id: BigInt(id) });
  return { success: response.deleted, id: Number(response.id) };
}

export async function deleteFeedbackBulk(ids: number[]): Promise<{ success: boolean; deleted: number }> {
  const response = await feedbackClient.deleteFeedbackBulk({ ids: ids.map((id) => BigInt(id)) });
  return { success: true, deleted: Number(response.deleted) };
}

File: scenarios/landing-page-business-suite/api/handlers/metrics/feedback.go
package metricshttp

import (
	"net/http"
	"strings"

	metrics "landing-page-business-suite-api/internal/metrics"
)

// seam: FeedbackNotifier delivers feedback notifications after persistence.
// Production is wired by the root composition package; tests use
// mocks.FakeFeedbackNotifier.
type FeedbackNotifier interface {
	Notify(*metrics.FeedbackRequest)
}

func (d Dependencies) CreateFeedback(svc metrics.FeedbackServicer, notifier FeedbackNotifier) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var input metrics.CreateFeedbackInput
		if !d.DecodeJSON(w, r, &input) {
			return
		}
		if strings.TrimSpace(input.Email) == "" {
			d.WriteErrorType(w, http.StatusBadRequest, "Email address is required.", "validation")
			return
		}
		if strings.TrimSpace(input.Subject) == "" {
			d.WriteErrorType(w, http.StatusBadRequest, "Subject is required.", "validation")
			return
		}
		if strings.TrimSpace(input.Message) == "" {
			d.WriteErrorType(w, http.StatusBadRequest, "Message is required.", "validation")
			return
		}
		if input.Type != "refund" && input.Type != "bug" && input.Type != "feature" && input.Type != "general" {
			input.Type = "general"
		}
		feedback, err := svc.Create(r.Context(), &input)
		if err != nil {
			d.LogError("feedback_create_failed", map[string]interface{}{"error": err.Error()})
			d.WriteErrorType(w, http.StatusInternalServerError, "Failed to submit feedback. Please try again.", "server_error")
			return
		}
		d.LogError("feedback_created", map[string]interface{}{"id": feedback.ID, "type": feedback.Type, "email": feedback.Email})
		notifier.Notify(feedback)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		if err := d.WriteJSON(w, map[string]interface{}{"success": true, "id": feedback.ID}); err != nil {
			d.LogError("feedback_response_encode_failed", map[string]interface{}{"error": err.Error()})
		}
	}
}

func (d Dependencies) ListFeedback(svc metrics.FeedbackServicer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		requests, err := svc.List(r.Context(), r.URL.Query().Get("status"))
		if err != nil {
			d.LogError("feedback_list_failed", map[string]interface{}{"error": err.Error()})
			d.WriteErrorType(w, http.StatusInternalServerError, "Failed to retrieve feedback list. Please try again.", "server_error")
			return
		}
		if requests == nil {
			requests = []metrics.FeedbackRequest{}
		}
		w.Header().Set("Content-Type", "application/json")
		if err := d.WriteJSON(w, requests); err != nil {
			d.LogError("feedback_list_encode_failed", map[string]interface{}{"error": err.Error()})
		}
	}
}

func (d Dependencies) GetFeedback(svc metrics.FeedbackServicer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, ok := d.PathInt(w, r, "id")
		if !ok {
			return
		}
		feedback, err := svc.GetByID(r.Context(), id)
		if err != nil {
			d.LogError("feedback_get_failed", map[string]interface{}{"id": id, "error": err.Error()})
			d.WriteErrorType(w, http.StatusNotFound, "Feedback not found.", "not_found")
			return
		}
		w.Header().Set("Content-Type", "application/json")
		if err := d.WriteJSON(w, feedback); err != nil {
			d.LogError("feedback_get_encode_failed", map[string]interface{}{"error": err.Error()})
		}
	}
}

func (d Dependencies) UpdateFeedbackStatus(svc metrics.FeedbackServicer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, ok := d.PathInt(w, r, "id")
		if !ok {
			return
		}
		var input struct {
			Status string `json:"status"`
		}
		if !d.DecodeJSON(w, r, &input) {
			return
		}
		if !validFeedbackStatus(input.Status) {
			d.WriteErrorType(w, http.StatusBadRequest, "Invalid status. Use: pending, in_progress, resolved, or rejected.", "validation")
			return
		}
		feedback, err := svc.UpdateStatus(r.Context(), id, input.Status)
		if err != nil {
			d.LogError("feedback_update_status_failed", map[string]interface{}{"id": id, "error": err.Error()})
			d.WriteErrorType(w, http.StatusInternalServerError, "Failed to update feedback status. Please try again.", "server_error")
			return
		}
		d.LogError("feedback_status_updated", map[string]interface{}{"id": id, "status": input.Status})
		w.Header().Set("Content-Type", "application/json")
		if err := d.WriteJSON(w, feedback); err != nil {
			d.LogError("feedback_status_encode_failed", map[string]interface{}{"error": err.Error()})
		}
	}
}

func (d Dependencies) DeleteFeedback(svc metrics.FeedbackServicer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, ok := d.PathInt(w, r, "id")
		if !ok {
			return
		}
		if err := svc.Delete(r.Context(), id); err != nil {
			d.LogError("feedback_delete_failed", map[string]interface{}{"id": id, "error": err.Error()})
			d.WriteErrorType(w, http.StatusInternalServerError, "Failed to delete feedback. Please try again.", "server_error")
			return
		}
		d.LogError("feedback_deleted", map[string]interface{}{"id": id})
		w.Header().Set("Content-Type", "application/json")
		if err := d.WriteJSON(w, map[string]interface{}{"success": true, "id": id}); err != nil {
			d.LogError("feedback_delete_encode_failed", map[string]interface{}{"error": err.Error()})
		}
	}
}

func (d Dependencies) DeleteFeedbackBulk(svc metrics.FeedbackServicer) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var input struct {
			IDs []int `json:"ids"`
		}
		if !d.DecodeJSON(w, r, &input) {
			return
		}
		if len(input.IDs) == 0 {
			d.WriteErrorType(w, http.StatusBadRequest, "No feedback IDs provided.", "validation")
			return
		}
		deleted, err := svc.DeleteBulk(r.Context(), input.IDs)
		if err != nil {
			d.LogError("feedback_bulk_delete_failed", map[string]interface{}{"ids": input.IDs, "error": err.Error()})
			d.WriteErrorType(w, http.StatusInternalServerError, "Failed to delete feedback items. Please try again.", "server_error")
			return
		}
		d.LogError("feedback_bulk_deleted", map[string]interface{}{"ids": input.IDs, "deleted": deleted})
		w.Header().Set("Content-Type", "application/json")
		if err := d.WriteJSON(w, map[string]interface{}{"success": true, "deleted": deleted}); err != nil {
			d.LogError("feedback_bulk_delete_encode_failed", map[string]interface{}{"error": err.Error()})
		}
	}
}

func validFeedbackStatus(status string) bool {
	return status == "pending" || status == "in_progress" || status == "resolved" || status == "rejected"
}
